miden-crypto/src/merkle/sparse_path.rs

```diff
@@ -0,0 +1,813 @@
+    pub fn iter(&self) -> impl ExactSizeIterator<Item = RpoDigest> {
+        self.into_iter()
+    }
+
+    // PRIVATE HELPERS
+    // ============================================================================================
+
+    const fn bitmask_for_depth(node_depth: NonZero<u8>) -> u64 {
+        // - 1 because paths do not include the root.
+        1 << (node_depth.get() - 1)
+    }
+
+    const fn is_depth_empty(&self, node_depth: NonZero<u8>) -> bool {
+        (self.empty_nodes_mask & Self::bitmask_for_depth(node_depth)) != 0
+    }
+
+    fn get_nonempty_index(&self, node_depth: NonZero<u8>) -> usize {
+        let bit_index = node_depth.get() - 1;
+        let without_shallower = self.empty_nodes_mask >> bit_index;
+        let empty_deeper = without_shallower.count_ones() as usize;
+        // The vec index we would use if we didn't have any empty nodes to account for...
+        let normal_index = (self.depth() - node_depth.get()) as usize;
+        // subtracted by the number of empty nodes that are deeper than us.
+        normal_index - empty_deeper
+    }
+}
+
+// SERIALIZATION
+// ================================================================================================
+
+impl Serializable for SparseMerklePath {
+    fn write_into<W: winter_utils::ByteWriter>(&self, target: &mut W) {
+        target.write_u8(self.depth());
+        target.write_u64(self.empty_nodes_mask);
+        target.write_many(&self.nodes);
+    }
+}
+
+impl Deserializable for SparseMerklePath {
+    fn read_from<R: winter_utils::ByteReader>(
+        source: &mut R,
+    ) -> Result<Self, DeserializationError> {
+        let depth = source.read_u8()?;
+        let empty_nodes_mask = source.read_u64()?;
+        let count = depth as u32 - empty_nodes_mask.count_ones();
+        let nodes = source.read_many::<RpoDigest>(count as usize)?;
+        Ok(Self { empty_nodes_mask, nodes })
+    }
+}
+
+// CONVERSIONS
+// ================================================================================================
+
+impl From<SparseMerklePath> for MerklePath {
+    fn from(sparse_path: SparseMerklePath) -> Self {
+        MerklePath::from_iter(sparse_path)
+    }
+}
+
+/// # Errors
+///
+/// This conversion returns [MerkleError::DepthTooBig] if the path length is greater than
+/// [`SMT_MAX_DEPTH`].
+impl TryFrom<MerklePath> for SparseMerklePath {
+    type Error = MerkleError;
+
+    fn try_from(path: MerklePath) -> Result<Self, MerkleError> {
+        SparseMerklePath::from_sized_iter(path)
+    }
+}
+
+impl From<SparseMerklePath> for Vec<RpoDigest> {
+    fn from(path: SparseMerklePath) -> Self {
+        Vec::from_iter(path)
+    }
+}
+
+// ITERATORS
+// ================================================================================================
+
+/// Contructs a [SparseMerklePath] out of an iterator of optional nodes, where `None` indicates an
+/// empty node.
+impl FromIterator<(NonZero<u8>, Option<RpoDigest>)> for SparseMerklePath {
```

Comment:
I don't think we can implement this without allowing users of the type to violate the guarantee that the depth never exceeds SMT_MAX_DEPTH. For example, this test currently passes:

```rust
#[test]
fn test_from_iter() {
    let tree = make_smt(8);

    let (control_path, _) = tree.open(&tree.entries().next().unwrap().0).into_parts();

    let path_depth_iter = path_depth_iter(tree.depth()).into_iter();
    let chain = [(NonZero::new(64u8).unwrap(), Some(control_path[0]))];
    let zipped = path_depth_iter.zip(control_path.into_iter().map(Some));
    let path = SparseMerklePath::from_iter(zipped.chain(chain));

    assert_eq!(path.depth(), 65);
}
```

So I think this should be an inherent, private method only used from the other safe constructors.

Reply:
Agreed that it should be a private method. Since it's only called from `from_iter_with_depth`, I internalized it there and made the whole thing private (as per my other comment).

As for the capacity, we don't know the Vec size until we consume `tree_depth` elements of the iterator and know how many empty elements there are. We don't store empty nodes.